Reject replies when author is blocked by thread owner

MrPetovan opened the pull request:
Closes #9464 

This change will only apply to future messages.

I was able to test that a public reply from a blocked Mastodon contact was rejected twice:
- Once as "Author is blocked by user" for the private copy of the item that should have appeared in my network feed (existing behavior)
- Once as "Author is blocked by parent user" for the public copy of the item that should have appeared in the community pages for example. (new behavior)

MrPetovan committed:
Remove obsolete references to item.parent-uri

annando commented on src/Model/Item.php:
```diff
@@ -1445,7 +1445,24 @@ public static function isValid(array $item)
-		
+
+		$toplevelOwnerId = null;
+		if ($item['parent-uri'] != $item['uri']) {
```
Please use the `gravity` to check for the parent. I want to unify this check that way.

MrPetovan replied:
Ok, I'll update the other instances in the same method.

annando commented on src/Model/Item.php:
```diff
@@ -1445,7 +1445,24 @@ public static function isValid(array $item)
+		$toplevelOwnerId = null;
+		if ($item['parent-uri'] != $item['uri']) {
+			$topLevelUriId   = ItemURI::getIdByURI($item['parent-uri']);
+			$toplevelItem    = self::selectFirst(['owner-link'], ['uri-id' => $topLevelUriId]);
```
You could simply check for a parent where `wall` is set to `true`. In the `uid` field of that post you would have the original thread starter.

MrPetovan replied:
Should that check be recursive, like `getParentData`?

annando replied:
There is no need for recursion. When you have got the real `parent-uri-id` then you can make a single query for a post with the `uri-id` set to the `parent-uri-id` and `wall` set to `true`. The `uid` is then the correct user.

annando commented on src/Model/Item.php:
```diff
@@ -1445,7 +1445,24 @@ public static function isValid(array $item)
+			Logger::notice('Owner is blocked by parent user', ['owner-link' => $item['owner-link'], 'uid' => $toplevelOwnerId, 'item-uri' => $item['uri']]);
+			return false;
+		}
+
```
Please also check for the `causer-id`.

MrPetovan replied:
The causer is currently only blocked by the user if the item is top-level, but the new check is specifically about replies, so I didn't think it was relevant.

annando replied:
A check for the causer wouldn't hurt. And I would not exclude some situation where the causer could contain a blocked person as well.

annando reviewed (changes requested):
Please have a look at the suggestions. But also please move your code into the function `getParentData` or after it had been called.

At the time that `isvalid` is called, we only know the thread parent - but this needn't to be the thread starter.

MrPetovan commented:
> Please have a look at the suggestions. But also please move your code into the function `getParentData` or after it had been called.

Isn't it a problem that it would be outside the `isValid()` method?

annando commented:
I think it would be great to have a separate function for this.

MrPetovan commented:
After looking into it:
- The `isValid` is used elsewhere, so it makes sense to check in the same method
- In the `Item::insert()` method, just after the `isValid()` call, there's a `getParentData` call which is creating a contact relation record. We need to abort the process before this call, so `isValid()` is the best place.

Now I'm pretty upset because `parent-uri` is used for the direct parent before the `getParentData` call and for the top-level parent after, which is awful, confusing and wasted my time in this PR. I'll likely have to rewrite this critical part of the code that I can't reliably test.

annando commented:
The check could easily be inside of the `getParentData` function. The second usage of `isValid` is only in the feeds (and there only for the date check). There we don't have any reason checking for this situation since the feed only creates toplevel posts and no comments.

MrPetovan commented:
But `getParentData` already does way much than the name implies (by creating the Contact relationship) so it would still not be in the right place and I still have to rewrite that part.

annando commented:
Well, the function is performing stuff concerning the parent post. So from my point of view it is a good choice. Also the function already has got some return value that could be used so that the item wasn't created. I, personally, would create some `isValidParent` function that was called by that function. And this can happen before the contact relation stuff.

MrPetovan commented:
Indeed, but it still doesn't fix the `parent-uri` field being used for two very distinct purposes.

annando commented:
The function contains the correct parent item array. And this whole `parent-uri` and `thr-parent` stuff is only mixed in the small part of the `Item::insert` function until the parent data is fetched.

MrPetovan commented:
Indeed, it should't take much to rewrite then, but I'd rather not have to feel I have to do it before submitting this PR. Oh well.

annando commented:
I just realized that the code here: https://github.com/friendica/friendica/blob/develop/src/Model/Item.php#L1408-L1447 (only the user checks) could be moved into a separate function that was called with the item array and the user id. Then the new function could be called from `isValid` and from `getParentData`.

Edit: The `parent-uri` and `thr-parent` has to be changed from everywhere where `Item::insert` is called. That's one of the reasons why I haven't changed it by now.

MrPetovan commented:
> I just realized that the code here: https://github.com/friendica/friendica/blob/develop/src/Model/Item.php#L1408-L1447 (only the user checks) could be moved into a separate function that was called with the item array and the user id. Then the new function could be called from `isValid` and from `getParentData`.

Why would it be called from `getParentData`? With what data?

annando commented:
In `getParentData` we do have the correct top level parent array. So we can check for an item with the same `uri-id` like the parent array but with `wall` enabled. The `uid` of that array can then be used.

MrPetovan commented:
Ah, I see, thanks.

MrPetovan commented:
https://github.com/friendica/friendica/blob/7fcaa1f23908dcdde48103e87aa975dd1e851017/mod/item.php#L608-L609

Guess who's changing this?

annando commented:
You had to check and change around 50 places in the code when changing this.

MrPetovan commented:
Yes, I'm knee-deep in this process, thank you for the support.

MrPetovan commented:
During my tests, I discovered that we transmit AP messages even to blocked contacts. Whoopsie. @tiotrom was wondering why his annoying follower kept commenting as easily after he blocked them, and well, now we know why.

MrPetovan committed:
Use item.thr-parent as expected in Model\Item::insert()

- Rework Model\Item::getTopLevelParent
- Backward compatibility with item.parent-uri is ensured

MrPetovan committed:
Replace confusing uses of item.parent-uri with expected item.thr-parent

MrPetovan committed:
Clarify item.parent-uri use in database field comment

MrPetovan committed:
Reject replies when author is blocked by thread owner in Model\Item::insert

- Move user-level item permission to Model\Item::isAllowedByUser
- Add user-level check for comments on top-level item

MrPetovan commented:
Here you go, @annando, take your time, I sure took mine. 🥱

MrPetovan commented:
I ended up not replacing URI checks with gravity because I'm not sure it's always exactly the same check. I don't know what GRAVITY_ACTIVITY is for example. If there was only two values for gravity, I would have made the change, but this is out-of-scope and out-of-my-depth.

annando commented:
`GRAVITY_PARENT` is **always** the parent.
`GRAVITY_COMMENT` is always a comment.
`GRAVITY_ACTIVITY` is used for likes, dislikes, announces, ...

After the function `getGravity` is called by the `insert` function, these values are 100% reliable.

annando commented:
Please also have a look for the corresponding fields in the addons where items are created.

MrPetovan commented:
Ha, yes, I have been modifying them but I didn't submit the change. Still, with the backward compatibility feature, it shouldn't break even if this PR is merged first/only.

> `GRAVITY_PARENT` is **always** the parent.

The direct parent or the top-level parent?

Edit: Nevermind, it is the top-level parent.

MrPetovan committed:
Replace uri fields conditions by gravity condition in Model\Item::insert

MrPetovan commented:
I'm having an issue on my dev node on this branch, submitting an item yield "Item wasn't stored." 500 error without any log entry.

But submitting it again seems to go through somehow?

MrPetovan committed:
Separate $parent_item and $toplevel_item in mod/item

MrPetovan committed:
Fix wrong variable use in Model\Item::getTopLevelParent

- It was preventing items at levels 3 and beyond to be inserted
- Logging for missing top level parent has been bumped to notice

MrPetovan commented:
I found my error, I used the wrong variable in `Item::getTopLevelParent` which prevented items from level 3 and more to be inserted. It looks good now.

MrPetovan commented:
> ```
> 1) Friendica\Test\legacy\ApiTest::testApiStatusesMediap
> Friendica\Database\DatabaseException: Column 'parent' cannot be null
> 
> /home/runner/work/friendica/friendica/src/Database/Database.php:756
> /home/runner/work/friendica/friendica/src/Database/Database.php:1000
> /home/runner/work/friendica/friendica/src/Database/DBA.php:302
> /home/runner/work/friendica/friendica/src/Model/Item.php:1900
> /home/runner/work/friendica/friendica/mod/item.php:724
> /home/runner/work/friendica/friendica/include/api.php:1045
> /home/runner/work/friendica/friendica/tests/legacy/ApiTest.php:1215
> ```


What now? I didn't touch this field, it is unset a couple times before the actual insertion (lines 1737 and 1779), and then it is added back to the item with an update, what is going on?

MrPetovan commented:
So the `parent` may need to be set after the item insertion because top level posts have their own id as parent, but it still doesn't explain why it was working before and now it doesn't.

I see that the default value for `item.parent` is 0, which means that a missing field value wouldn't abort the INSERT, but a present value of NULL would.

MrPetovan committed:
Fix null value for item.parent column

MrPetovan commented:
This time it went through, apparently unsetting an array key holding `null` doesn't remove the key itself?

MrPetovan committed:
Ensure the parent field isn't set during Item insertion

- Avoid a database error if a null value is provided

annando commented:
Shouldn't you change https://github.com/friendica/friendica/blob/20f88354ebbdd8c6930af11b5dc4f8edf7f468cf/include/api.php#L1109 as well?

MrPetovan commented:
No, this doesn't need to change, we were already storing the value of this REQUEST parameter in the local `$thr_parent_uri` variable in `mod/item.php:104`. I *could* change it but it's already a big enough PR.

annando commented:
Here `thr-parent` should be used, I think: https://github.com/friendica/friendica/blob/develop/src/Worker/OnePoll.php#L586-L588

MrPetovan commented:
You're right. It still would work thanks to the backward compatibility feature, but it would be nicer.

annando commented on src/Protocol/Feed.php:
```diff
@@ -354,7 +354,7 @@ public static function import($xml, array $importer = [], array $contact = [])
 			$item["plink"] = DI::httpRequest()->finalUrl($item["plink"]);
 
-			$item["parent-uri"] = $item["uri"];
+			$item["thr-parent"] = $item["uri"];
```
Is this really needed?

MrPetovan replied:
No, this was from before I started to remove self-assigned `parent-uri` and `thr-parent` before passing to `Item::insert`.

annando commented on src/Protocol/Feed.php:
```diff
@@ -1108,9 +1109,8 @@ private static function entryFooter(DOMDocument $doc, $entry, array $item, array
-			$parent_item = (($item['thr-parent']) ? $item['thr-parent'] : $item['parent-uri']);
 
-			$thrparent = Item::selectFirst(['guid', 'author-link', 'owner-link', 'plink'], ['uid' => $owner["uid"], 'uri' => $parent_item]);
+			$thrparent = Item::selectFirst(['guid', 'author-link', 'owner-link', 'plink'], ['uid' => $owner["uid"], 'uri' => $item['thr-parent']]);
```
On really old posts the field `thr-parent` had been empty. So I guess we need an update job that fills `thr-parent` with the value of `parent-uri` when it's empty.

annando replied:
I just checked. Indeed I have got around 16,000 posts from the year 2012 where this is the case.

annando commented on src/Protocol/OStatus.php:
```diff
@@ -523,9 +523,9 @@ private static function process($xml, array $importer, array &$contact = null, &
 				} else {
 					// But we will only import complete threads
-					$valid = Item::exists(['uid' => $importer["uid"], 'uri' => self::$itemlist[0]['parent-uri']]);
+					$valid = Item::exists(['uid' => $importer["uid"], 'uri' => self::$itemlist[0]['thr-parent']]);
```
We are checking for the existence of the complete thread, so we have to look for the `parent-uri`.

MrPetovan commented:
> Here `thr-parent` should be used, I think: https://github.com/friendica/friendica/blob/develop/src/Worker/OnePoll.php#L586-L588

Actually it's handled once here: https://github.com/friendica/friendica/blob/20f88354ebbdd8c6930af11b5dc4f8edf7f468cf/src/Worker/OnePoll.php#L625-L628

annando commented on src/Protocol/DFRN.php:
```diff
@@ -438,7 +438,7 @@ public static function mail($item, $owner)
 
 		XML::addElement($doc, $mail, "dfrn:id", $item['uri']);
-		XML::addElement($doc, $mail, "dfrn:in-reply-to", $item['parent-uri']);
+		XML::addElement($doc, $mail, "dfrn:in-reply-to", $item['thr-parent']);
```
This is the "mail" part. I don't think that we do have a "thr-parent" here.

annando commented on src/Protocol/DFRN.php:
```diff
@@ -956,10 +956,9 @@ private static function entry(DOMDocument $doc, $type, array $item, array $owner
+			$parent = Item::selectFirst(['guid', 'plink'], ['uri' => $item['thr-parent'], 'uid' => $item['uid']]);
 			if (DBA::isResult($parent)) {
-				$attributes = ["ref" => $parent_item, "type" => "text/html",
+				$attributes = ["ref" => $item['thr-parent'], "type" => "text/html",
```
Same like in the feeds: We should create an update function to fill empty `thr-parent` fields.

annando reviewed (changes requested):
I will review the `Item` model class later, maybe tomorrow. Here are my changes so far.

annando commented:
Concerning the `OnePoll` class I realized that we always fetch the parent here. Means that we don't handle nested replies. That's what I meant.

MrPetovan commented:
I understand, but since `Item::insert` does it no matter what, there's no reason to set `parent-url` in the parameter array, it will be rewritten.

MrPetovan committed:
Remove references to item.parent-uri in Worker\OnePoll

MrPetovan committed:
Clarify parameter type in DFRN::mail

MrPetovan committed:
Remove unnecessary data array assignment in Protocol\Feed

MrPetovan committed:
Revert wrong item.thr-parent field usage in Protocol\OStatus

MrPetovan committed:
[Database version 1375] Add update method to populate missing item.thr-parent values

MrPetovan commented:
All the feedback so far has been addressed, including the suggested change to `Protocol\OnePoll` outside of the review.

annando commented on update.php:
```diff
@@ -747,4 +747,14 @@ function pre_update_1365()
+
+function update_1374()
+{
+	if (!DBA::e("UPDATE `item` SET `thr-parent` = `parent-uri` WHERE `thr-parent` = ''")) {
```
Please also set the `thr-parent-id` with the `parent-uri-id` for these cases.

MrPetovan replied:
Done.

annando commented on src/Model/Item.php:
```diff
@@ -1497,88 +1480,50 @@ private static function getDuplicateID(array $item)
+			'parent-uri' => $parent['parent-uri'],
+			'uid' => $parent['uid']];
+		// We select wall = 1 in priority for top level permission checks
+		$params = ['order' => ['wall' => true]];
```
This doesn't look right to me. We do have a parent item for this user and additionally we **might** have a parent where `wall` is true. These will very often be two different item arrays.

MrPetovan replied:
I don't understand what I should change then.

Edit: Wait, is it because `wall` would only be `true` for `uid = <owner-id`? So we probably should look for the parent item (which can be `uid = 0` so `wall = false` by definition) and the parent item local author separately.

MrPetovan replied:
I now created a separate query to retrieve the local parent top level item if any.

annando commented on src/Model/Item.php:
```diff
@@ -1497,88 +1480,50 @@ private static function getDuplicateID(array $item)
-			} else {
-				$item['wall'] = false;
-			}
+		if ($parent['uri'] == $parent['parent-uri']) {
```
When this is the case, you currently aren't checking for `isAllowedByUser`.

MrPetovan replied:
Yes, in `isValid` above.

annando replied:
But that only checks for the user of this parent. This doesn't work if there is a local post from another user with this parent-uri. This checks needs to be applied here as well: https://github.com/friendica/friendica/pull/9515/files#diff-785d24b46fdf749363656f33208f941ea0eca809c7c76452180a0bfc16df68a3R1519-R1522

MrPetovan replied:
Aaaah, I see what you mean now. So I need to perform this check outside of the `getTopLevelParentData` method.

MrPetovan replied:
Done.

MrPetovan committed:
Retrieve local top level parent item separately to check permissions in Model\Item::getTopLevelParentData

MrPetovan committed:
Move top-level permission check outside of Model\Item::getTopLevelParentData

- It wasn't checked when the direct parent was also the top-level parent

annando commented:
Can you add the updated `database.sql`?

annando commented on src/Model/Item.php:
```diff
@@ -1740,6 +1676,68 @@ public static function insert($item, $notify = false, $dontcache = false)
+			$condition = ['uri' => $toplevel_parent['uri'], 'wall' => true];
+			$localTopLevelParent = self::selectFirst(['uid'], $condition);
+			if (!empty($localTopLevelParent['uid']) && !self::isAllowedByUser($item, $localTopLevelParent['uid'])) {
+				return 0;
```
We should add some log message here, I think.

MrPetovan replied:
Log messages are sent by `isAllowedByUser` in all the `return false` cases.

I know it feels weird, I've wondered this multiple times myself but it's already handled.

annando replied:
But in this log messages we can't see if a post is rejected because the current user or the thread owner had blocked the author. I would prefer such a log message especially to be able to check if this works at all. It's good for support requests, I think.

MrPetovan replied:
The particular user id is included in the logger context.

MrPetovan committed:
Update database.sql with the latest structure changes

MrPetovan commented:
> Can you add the updated `database.sql`?

Done.

AlfredSK commented:
After pulling the update with this PR I'm literally getting more than 100 admin notification mails about the database structure update. And they're still being created and sent out. 114 and counting...

"The friendica database was successfully updated from 1374 to 1375."

MrPetovan commented:
I've got 16 myself on my own node. For a field comment change!

AlfredSK commented:
Pffff. 16. I got 239 now and more to come. 😜

AlfredSK commented:
Ok. It has stopped after 5.5 hours. Final box score: 650 mails.

MrPetovan commented:
Oof.

AlfredSK commented:
@MrPetovan Can you please have a look at https://github.com/friendica/friendica/issues/9252#issuecomment-729171051 ?
